File: modulo7-Ordenação/quick-sort.py
import math 
from random import random as aleatorio

def main():
    colecao = [8, 1, 10, 5, 3, 6]
    
    #colecao = ['Mari', 'Ana', 'Pedrinho', 'Rogerio', 'Zenóbia', 'Ana Leticia']
    
    colecao = ['Mari', 'Ana', 'Pedrinhu', 'Rogerio', 'Zenóbian', 'Ana Letíciab']
    
    print(colecao)
    
    ##Coleção de Números:
    #colecao_ordenada = quick_sort(colecao)  #[1, 3, 5, 6, 8, 10] & ['Ana', 'Ana Leticia', 'Mari', 'Pedrinho', 'Rogerio', 'Zenóbia']
    #colecao_ordenada = quick_sort(colecao, reverse=True)  #[10, 8, 6, 5, 3, 1]
    
    
    ##Coleção de Nomes:
    #colecao_ordenada = quick_sort(colecao, chave=len)  #['Ana', 'Mari', 'Zenóbia', 'Rogerio', 'Pedrinho', 'Ana Leticia']
    #colecao_ordenada = quick_sort(colecao, chave=len, reverse=True)  #['Ana Leticia', 'Pedrinho', 'Rogerio', 'Zenóbia', 'Mari', 'Ana'] 
    colecao_ordenada = quick_sort(colecao, chave=ultima_letra)  #['Ana', 'Ana Letíciab', 'Mari', 'Zenóbian', 'Rogerio', 'Pedrinhu']
    
    print(colecao_ordenada) 
    

def ultima_letra(nome):
    return nome[len(nome) - 1]
    
    
def quick_sort(vetor, chave=lambda x:x, reverse=False):
    if len(vetor) <= 1:
        return vetor
    
    random_index = math.floor(aleatorio() * len(vetor))
    pivot = vetor[random_index]
    
    vetor.remove(pivot)
    
    if reverse:
        left = filtro_colecao(vetor, lambda x: chave(x) > chave(pivot))
        rigth = filtro_colecao(vetor, lambda x: chave(x) <= chave(pivot))
    else:
        left = filtro_colecao(vetor, lambda x: chave(x) <= chave(pivot))
        rigth = filtro_colecao(vetor, lambda x: chave(x) > chave(pivot)) 
        
    return quick_sort(left, chave, reverse) + [pivot] + quick_sort(rigth, chave, reverse)
       

def filtro_colecao(colecao, funcao_criterio):
    colecao_filtrada = []
    
    for item in colecao:
        if funcao_criterio(item):
            colecao_filtrada.append(item)
    
    return colecao_filtrada            
    

''' def mapear_colecao(colecao, funcao_mapeadora):
    nova_colecao = []
    
    for item in colecao:
        nova_colecao.append(funcao_mapeadora(item))

    return nova_colecao '''
    
    
main()

File: modulo7-Ordenação/quick-sort.js
function main(){
    //const colecao = [8, 1, 10, 5, 3, 6]

    const colecao = ['Mari', 'Ana', 'Pedrinho','Rogerio', 'Zenóbia', 'Ana Letícia']
    //const colecao = ['Mari', 'Ana', 'Pedrinhu', 'Rogerio', 'Zenóbian', 'Ana Letíciab']

    console.log(colecao)

    // Coleção de Números
    //const colecao_ordenada = quicksort(colecao)
    // const colecao_ordenada = quicksort(colecao, {reverse: true})


    // Colecao de Nomes
    //const colecao_ordenada = quicksort(colecao, {chave: len}) //Ordem de menor a maior nomes
    //const colecao_ordenada = quicksort(colecao, {chave: len, reverse: true}) //Ordem do último nome ao primeiro
    const colecao_ordenada = quicksort(colecao, {chave: ultima_letra}) //Ordena pela ultima letra de cada nome 

    console.log(colecao_ordenada)
}

function len(vetor){
    return vetor.length
}

function ultima_letra(nome){
    return nome[nome.length - 1]
}

export function quicksort(vetor, {chave = x => x, reverse = false} = {}){

    if (vetor.length <= 1)
        return vetor

    const random_index = Math.floor(Math.random() * vetor.length)
    const pivot = vetor[random_index]
    // sorry: using method (implemente remover_item_colecao(colecao, item))
    vetor.splice(random_index, 1)

    let left, right

    if (reverse){
        left = filtro_colecao(vetor,  x => chave(x) > chave(pivot))
        right = filtro_colecao(vetor, x => chave(x) <= chave(pivot))
    }else{
        left = filtro_colecao(vetor,  x => chave(x) <= chave(pivot))
        right = filtro_colecao(vetor, x => chave(x) > chave(pivot))
    }

    return quicksort(left, {chave, reverse}).concat([pivot]).concat(quicksort(right, {chave, reverse}))
}

// local: utils
function filtro_colecao(colecao, funcao_criterio){
    const colecao_filtrada = []

    for (const item of colecao){
        if (funcao_criterio(item)){
            colecao_filtrada.push(item)
        }
    }

    return colecao_filtrada
}


main()